plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java

```diff
@@ -583,31 +576,25 @@ private class AddFileEntryExtractor
     {
         @Nullable
         @Override
-        public DeltaLakeTransactionLogEntry getEntry(ConnectorSession session, int pagePosition, Block... blocks)
+        public DeltaLakeTransactionLogEntry getEntry(ConnectorSession session, int pagePosition, Block addBlock)
         {
-            checkState(blocks.length == getRequiredChannels(), "Unexpected amount of blocks: %s", blocks.length);
-            Block addBlock = blocks[0];
-            Block addPartitionValuesBlock = blocks[1];
             log.debug("Building add entry from %s pagePosition %d", addBlock, pagePosition);
             if (addBlock.isNull(pagePosition)) {
                 return null;
             }
 
-            checkState(!addPartitionValuesBlock.isNull(pagePosition), "Inconsistent blocks provided while building the add file entry");
-            SqlRow addPartitionValuesRow = getRow(addPartitionValuesBlock, pagePosition);
-            CheckpointFieldReader addPartitionValuesReader = new CheckpointFieldReader(session, addPartitionValuesRow, addPartitionValuesType.orElseThrow());
-            Map<String, String> partitionValues = addPartitionValuesReader.getMap(stringMap, "partitionValues");
-            Map<String, Optional<String>> canonicalPartitionValues = canonicalizePartitionValues(partitionValues);
-            if (!partitionConstraint.isAll() && !partitionMatchesPredicate(canonicalPartitionValues, partitionConstraint.getDomains().orElseThrow())) {
-                return null;
-            }
-
             // Materialize from Parquet the information needed to build the AddEntry instance
             addBlock = addBlock.getLoadedBlock();
             SqlRow addEntryRow = getRow(addBlock, pagePosition);
             log.debug("Block %s has %s fields", addBlock, addEntryRow.getFieldCount());
             CheckpointFieldReader addReader = new CheckpointFieldReader(session, addEntryRow, addType.orElseThrow());
 
+            Map<String, String> partitionValues = addReader.getMap(stringMap, "partitionValues");
+            Map<String, Optional<String>> canonicalPartitionValues = canonicalizePartitionValues(partitionValues);
+            if (!partitionConstraint.isAll() && !partitionMatchesPredicate(canonicalPartitionValues, partitionConstraint.getDomains().orElseThrow())) {
+                return null;
+            }
```

Comment:
I think I understand what you are saying, but no, this can't move and I don't think it would matter.  It cannot move because this code is using the `addReader` variable.  This variable is created in the line before this and it uses the `addEntryRow` variable which is read from the `addBlock`.  Or said another way, the previous block is creating all of the data used in this block.  As for why this doesn't matter, all of this work is to remove the whole concept of lazy blocks from Trino.  This means that when you have a Block is it always materialized.  The PR for that change is queued waiting for this PR.

Reply:
I get that we want to remove LazyBlock construct to simplify code, but does that mean that the feature itself is going away and there is going to be no other way to achieve the same outcome of lazy materialization ? (btw engines which don't have it, want to have it https://issues.apache.org/jira/browse/SPARK-42256).
As for this specific code, this was implemented as an optimization in https://github.com/trinodb/trino/pull/19795 and it is impactful for delta lake query planning.
fyi @findinpath @ebyhr

Reply:
@raunaqmorarka I did some more digging.  The [current code](https://github.com/trinodb/trino/blame/master/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java#L587-L598) always materializes the blocks.  It happens because the code is calling `addBlock.isNull` and to compute that the block must be loaded.

BTW this code is super common in our codebase.  Folks go out of there way to try to delay materialization.